### PR TITLE
Revise x86_64 inline asm: Fix for LDC v1.29+, and enable on Windows and with GDC too

### LINKED ISSUE #457: Program exited with code -11 on not Windows and ldc-1.29.0
When using `DLexer` with `sse42 = true` on not Windows and ldc-1.29.0, DLexer constructor has crashed with exited code -11.

Reproducing code is bellow.

```d
/+dub.sdl:
dependency "libdparse" version="~>0.19.1"
+/
import std;

void main()
{
    import dparse.lexer : LexerConfig, StringCache, DLexer;

    auto sourceCode = q{
        int x;
    };  

    LexerConfig config;
    auto cache = StringCache(StringCache.defaultBucketCount);
    auto lexer = DLexer(sourceCode, config, &cache); // Program exited with code -11 

    // this code reproduce the error too.
    // auto lexer = DLexer(sourceCode, config, &cache, true);

    // But this code does not reproduce the error
    //auto lexer = DLexer(sourceCode, config, &cache, false);
}
```

This error is not occurred with ldc-1.28.0.

Error location is `skip` function in`src/dparse/lexer.d`.

```d
asm pure nothrow @nogc
{
    naked;
    movdqu XMM1, [RDX];
    mov R10, constant;
    movq XMM2, R10;
    mov RAX, charsLength;
    mov RDX, 16;
    pcmpestri XMM2, XMM1, flags;
    add [RSI], RCX;
    add [RDI], RCX;
    ret;
}
```

### PR DESCRIPTION
**Pull request by kinke**
Resolves #457.

### CODE REVIEW

**Comment by kinke** [comment]
The GDC/GCC-style inline asm syntax is also inline-able with LDC, unlike DMD-style one. IIRC, LDC v1.22 added support for that syntax.

**Comment by WebFreak001** [comment]
@dd86k can you review this?

**Comment by kinke** [comment]
FWIW, here's the almost-identical GCC-style one for C++, supported by gcc and clang: [godbolt](https://cpp.godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGe1wAyeAyYAHI%2BAEaYxCBmAKykAA6oCoRODB7evnrJqY4CQSHhLFEx8baY9vkMQgRMxASZPn5cFVXptfUEhWGR0bEJCnUNTdmtQ109xaUDAJS2qF7EyOwcBACeiZhYVADUXgypwCHou7QCwGcX%2B8EEAGySAPoEJgDMVhoAgiZfXrcPz12iWQLC2Q2IeAgaEOBF2yAQ9V2ACoIusCJgFLMft8AOwfT67QlwgRDTCqRLEG6CAGw6ETQS7N4AEV2Zg0bLicTuAA4AJxvfFE4kwskUqn3J60hHEBQBRjAAgIRmvFmSAXYoV09Hkym3XZUAzABTKlkaCIaC1cC0W9VfDVEv7UyW7YgYry0F7ve2EpgKFje3YQANC8xmFioABu6AAjl4Q/FWoz4uY4qoWCwuIy4m4GCZs8GheGI9GhYSU69SEm4im0ywzFmc3m3AWicDQRiCBCqwkqzX04mU3266GW7sQEmzMzTGZ6xBXQp3QRZkSAPQrqsacdtsGdvAut0e3bBXYAJTcAA1K0RdlFdkMSNsj4JUIy7hoGD4TG/R%2BPQ/6Z4GKJohi8yruuKZcOOTDoOg87GqgexfmawEKEhd67kYo5ljO6ChoGXB3KBpa7GuVZmOO9BGIquwIa%2ByHoqhb67AgmDQU%2BuzngAspxia%2BkCII7l2wSJF4sLHtgTLnlhE7EHhUIknUgigaR5bjiwTCqLs3J0fC9SMRouzXucbG3C%2BADiyj7sAeCkmQhksQwHHcZOXzEfGZhMHJukynKVEIMp4HxJIFHytRx5cZxZiVnx24dkJDAiWJjnYJ8UmuW52FmHgckGkwRrLkKKnxHEv5vrlRpoXg6bbHgTDoj%2BE61lwoaVqGtaTgBJHrsg5wRKUj6utZtmoelhJYl6o37gQSyOfOi62riTLYhw8y0JwcS8H4HBaKQqCcM2ljWHeizLJgSavDwpAEJoK3zAA1iArxxAAdHEkjveyZiSGYOJsjO%2BicJIm03btnC8AoIAaFdN3zHAsBIGgoJ0NE5CUIjiTIzEPWGMAUgWnwdDojKlARJwl0RME9TrGTvCI2wggAPIMLQ1PcLwWDqUY4jbezeCug4eARhiNOkGSmDIKJqw7bclQi7QeARMQVMeFgIu7iwNPzOVCgAGp4JgADuDNbFtl38IIIhiOwUgyIIigqOoPOkLorQGEYKDWNY%2BgKxDkDzKgiTVBDHAALQM684OVOL1QuAw7ieM0/hx1MfQxK0uRpAIowtEkKSZwwKclP04xRwLAidCMCdjG00cdMM3TBL0Rdp7Y9fZ3oEwNIXMxcPMCgnSseidpgqw8Kt63A07e0cKo3J3CHDxwm7lxSM9Ghr4GB1WJYla4IQJDnYmHhI/QlLmBdsy8NdPOzPdj1cM9XAzmYDzfa8vKSA8OJ3ADHBA6QW0drT3BpDaGN9SBw0QCgVAJ8UYUHkrAmIuMuAVhoB6aIEMICkydhTZgxBWaXTpowAgTMWYgw5jjbmO18D80cELYOO0xYS3RCLGWa02be0VsrDAUsr4Qg1mzLWhpdb6yNibEW5thCiHEDbSR9s1Ag10FFZeHtDo73lhEX2EB/aB3SMHMOEca5lz8BAVw7dWiBEbtMYuuc8jpHMbY/O3cbF2FruXNuVcc6uOMRXBuRRU4dw8VkLx9dnFpz7gPa2V1XSjxWr/DaACQbT1nvPRewBkDIF2FwB%2BrxN6ex3rsPeRAz5mAupWY%2BGNT7nTMJfMBWhb6kAeq8V4z1mltPaR03%2B/9AG8GAbYUB196njw4GYSeQCwZ1NuqQIWMp0ggEkEAA)

**Comment by dd86k** [comment]
> @dd86k can you review this?

@WebFreak001: I don't mind testing this, but what would I be able to test, or should I expect from this PR, precisely?

**Comment by kinke** [comment]
The unittest runner (`dub test`) would fail for me without this change with LDC v1.29 on Linux (#457, obviously with a CPU with SSE 4.2), now works. I haven't tested it on Windows, nor with GDC.

**Comment by dd86k** [comment]
To resume: Run tests with this PR on a Linux system with SSE4.2 using LDC 1.29.0 and GDC? Optionally on Windows?

**Comment by kinke** [comment]
Yep, that should cover everything I think. - Note that LDC versions < 1.22 couldn't be used anymore to compile libdparse; that would require some more static if + enum business to use the DMD-style syntax for those.

**Comment by dd86k** [comment]
Cloned `https://github.com/kinke/libdparse asm` onto my machine.

Commit: `0.19.1+commit.1.g378e41e`

Tests using `dub test --compiler=ldc2` and `dub test --compiler=gdc` respectively:

- Linux x86_64 5.16.19
  - LDC 1.28.0: `8 modules passed unittests`
  - GDC 11.2.0: `All unit tests have been run successfully.`
- Windows 10 21H1 x86_64 10.19044.1645
  - LDC 1.28.1: `8 modules passed unittests`
- Windows 11 21H2 x86_64 10.22000.613
  - LDC 1.29.0: `8 modules passed unittests`
  - Note: VS2019 Build Tools + VS2022 Community installed

Didn't test GDC on Windows because last time I tried using GDC on Windows last year, it couldn't compile D code at all.

Hope this answers it for you! I'll be here if you need anything else.

## Processor info

Linux (host):
```
Name:        AuthenticAMD AMD Ryzen 9 5950X 16-Core Processor            
Identifier:  Family 0x19 Model 0x21 Stepping 0x0
Cores:       16 cores, 32 threads
Max. Memory: 256 TiB physical, 256 TiB virtual
Baseline:    x86-64-v3
Techs:       core-performance-boost htt
Extensions:  x87/fpu +f16c mmx extmmx amd64/x86-64 +lahf64 amd-v/vmx +svm=v1 aes-ni adx sha bmi1 bmi2
SSE:         sse sse2 sse3 ssse3 sse4.2 sse4a
AVX:         avx avx2
AMX:         None
Mitigations: ibpb ibrs ibrs_pref stibp stibp_on ssbd
Cache L1-D:   16x   32 KiB,  512 KiB total, si
Cache L1-I:   16x   32 KiB,  512 KiB total, si
Cache L2-U:   16x  512 KiB,    8 MiB total, si ci
Cache L3-U:    2x   32 MiB,   64 MiB total, si nwbv
```

Windows 10 (guest, should be same for 11):
```
Name:        AuthenticAMD AMD Ryzen 9 5950X 16-Core Processor
Identifier:  Family 0x19 Model 0x21 Stepping 0x0
Cores:       2 cores, 4 threads
Max. Memory: 256 TiB physical, 256 TiB virtual
Baseline:    x86-64
Techs:       htt
Extensions:  x87/fpu mmx extmmx amd64/x86-64 +lahf64 amd-v/vmx +svm=v1 aes-ni
SSE:         sse sse2 sse3 ssse3 sse4.2 sse4a
AVX:         avx avx2
AMX:         None
Mitigations:
ParaVirt.:   Hyper-V
Cache L1-D:    2x   32 KiB,   64 KiB total, si
Cache L1-I:    2x   32 KiB,   64 KiB total, si
Cache L2-U:    2x  512 KiB,    1 MiB total, si ci
Cache L3-U:    1x   32 MiB,   32 MiB total, si nwbv
```

File: src/dparse/lexer.d [hunk 1]
```diff
@@ -2198,61 +2193,85 @@ unittest
...
+            return result;
+        }
+    }
+
     /**
      * Skips between 0 and 16 bytes that match (or do not match) one of the
      * given $(B chars).
      */
-    void skip(bool matching, chars...)(const ubyte*, ulong*, ulong*) pure nothrow
+    void skip(bool matching, chars...)(const ubyte* bytes, ulong* pindex, ulong* pcolumn) pure nothrow
```
Comment (maxhaton) [inline src/dparse/lexer.d:2249]: Can these attributes not be inferred?

Comment (kinke) [reply src/dparse/lexer.d:2256]: Most likely, but I didn't want to mix in other refactorings (the pointers would have been gone first and replaced by refs obviously...).

**Comment by kinke** [comment]
[pushed a little extension to make LDC < v1.22 work again]

File: src/dparse/lexer.d (hunk 1 above)
Comment (maxhaton) [reply src/dparse/lexer.d:2256]: I think avoiding refactoring solely on principle is a good way to kill progress.

**Comment by maxhaton** [comment]
In some quick tests I did, this whitespace skipping approach doesn't seem to be particularly quick, I'm not convinced there's all that much point in keeping this at all.

File: src/dparse/lexer.d (hunk 1 above)
Comment (kinke) [reply src/dparse/lexer.d:2256]: I don't avoid it by principle, I decide from time to time. The point here is clearly asm, not tampering with a codebase I've seen first yesterday and don't plan to contribute to regularly.

**Review by RazvanN7 — APPROVED** [review]

**Comment by edi33416** [comment]
> In some quick tests I did, this whitespace skipping approach doesn't seem to be particularly quick, I'm not convinced there's all that much point in keeping this at all

I agree. While I like reading some assembly as much as the next person, I don't think we should use inline asm anymore.
I think this code will hinder better optimizations, such as using the `ymm` or `zmm` registers for the newer cpus.
I'm pretty sure that @dd86k 's 16 core would do better without the inline asm.

I believe that the inline asm should be removed in a future PR. This will prevent future issues like this one, and will enable the compiler to perform optimizations as the compiler and architectures evolve without requiring code patches.

**Comment by RazvanN7** [comment]
This PR: https://github.com/dlang-community/D-Scanner/pull/866 confirms that this fix is correct, however, we still need a new tag so that the dub test also passes. cc @WebFreak001

**Comment by WebFreak001** [comment]
ok will tag release

**Comment by RazvanN7** [comment]
@WebFreak001 Is there anything else that needs to be done for the dub registry to pick up the new tag or does it simply take a while before it gets updated? I see that the dub registry [1] still thinks that 0.19.1 is the latest tag.

[1] https://code.dlang.org/packages/libdparse

**Comment by WebFreak001** [comment]
ping @Hackerpilot 

otherwise wait for 0~8 hours

**Pull request merged**